File: go/problem37.go
package main

import (
	"fmt"
	"math"
)

func main() {
	fmt.Println(TruncatablePrimeSum(8))
}

func TruncatablePrimeSum(requiredNum int) int {

	count := 0
	sum := 0
	num := 11
	for count < requiredNum {
		if isPrime(num) && isTruncatablePrime(num) {
			count++
			sum += num
			// fmt.Println("Found a truncatable prime ", num)
		}
		num += 2
	}
	return sum
}

func isTruncatablePrime(num int) bool {

	for div := 10; div < num; div *= 10 {
		if !isPrime(int(num/div)) || !isPrime(num%div) {
			return false
		}
	}
	return true
}

func isPrime(num int) bool {

	if num <= 1 {
		return false
	}

	if num == 2 {
		return true
	}

	if num%2 == 0 {
		return false
	}

	for i := 3; i <= int(math.Sqrt(float64(num))); i += 2 {
		if num%i == 0 {
			return false
		}
	}

	return true
}

File: python/problem37.py
import math


def truncatablePrimeSum(requiredNum):
    count = 0
    sum = 0
    num = 11
    while count < requiredNum:
        if isPrime(num) and isTruncatablePrime(num):
            count += 1
            sum += num
        num += 2

    return sum


def isTruncatablePrime(num):
    div = 10
    while div < num:
        if not isPrime(int(num // div)) or not isPrime(num % div):
            return False
        div *= 10

    return True


def isPrime(num):
    if num <= 1:
        return False

    if num == 2:
        return True

    if num % 2 == 0:
        return False

    for i in range(3, int(math.sqrt(num)) + 1, 2):
        if num % i == 0:
            return False

    return True


if __name__ == "__main__":
    print(truncatablePrimeSum(8))
